hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkSampleWritesUtils.java

```diff
@@ -0,0 +1,143 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkSampleWritesUtils.class);
+
+  public static void overwriteRecordSizeEstimateIfNeeded(JavaSparkContext jsc, JavaRDD<HoodieRecord> records, HoodieWriteConfig writeConfig, String instantTime) {
+    if (!writeConfig.getBoolean(SAMPLE_WRITES_ENABLED)) {
+      LOG.debug("Skip overwriting record size estimate as it's disabled.");
+      return;
+    }
+    HoodieTableMetaClient metaClient = getMetaClient(jsc, writeConfig.getBasePath());
+    if (metaClient.isTimelineNonEmpty()) {
+      LOG.info("Skip overwriting record size estimate due to timeline is non-empty.");
+      return;
+    }
+    try {
+      Pair<Boolean, String> result = doSampleWrites(jsc, records, writeConfig, instantTime);
+      if (result.getLeft()) {
+        long avgSize = getAvgSizeFromSampleWrites(jsc, result.getRight());
+        LOG.info("Overwriting record size estimate to " + avgSize);
+        writeConfig.setValue(COPY_ON_WRITE_RECORD_SIZE_ESTIMATE, String.valueOf(avgSize));
+      }
+    } catch (IOException e) {
+      LOG.error(String.format("Not overwriting record size estimate for table %s due to error when doing sample writes.", writeConfig.getTableName()), e);
+    }
+  }
+
+  private static Pair<Boolean, String> doSampleWrites(JavaSparkContext jsc, JavaRDD<HoodieRecord> records, HoodieWriteConfig writeConfig, String instantTime) throws IOException {
+    long now = Instant.now().toEpochMilli();
+    Path basePath = new CachingPath(writeConfig.getBasePath(), SAMPLE_WRITES_FOLDER_PATH + Path.SEPARATOR + instantTime + Path.SEPARATOR + now);
+    final String sampleWritesBasePath = basePath.toString();
+    HoodieTableMetaClient.withPropertyBuilder()
+        .setTableType(HoodieTableType.COPY_ON_WRITE)
+        .setTableName(String.format("%s_samples_%s_%s", writeConfig.getTableName(), instantTime, now))
+        .setCDCEnabled(false)
+        .initTable(jsc.hadoopConfiguration(), sampleWritesBasePath);
+    HoodieWriteConfig sampleWriteConfig = HoodieWriteConfig.newBuilder()
+        .withProps(writeConfig.getProps())
+        .withPath(sampleWritesBasePath)
+        .withMetadataConfig(HoodieMetadataConfig.newBuilder().enable(false).build())
+        .withSampleWritesEnabled(false)
+        .withTableServicesEnabled(false)
+        .withSchemaEvolutionEnable(false)
+        .withBulkInsertParallelism(1)
+        .withAutoCommit(true)
+        .build();
+    try (SparkRDDWriteClient sampleWriteClient = new SparkRDDWriteClient(new HoodieSparkEngineContext(jsc), sampleWriteConfig, Option.empty())) {
+      int size = writeConfig.getIntOrDefault(SAMPLE_WRITES_SIZE);
+      List<HoodieRecord> samples = records.coalesce(1).take(size);
+      sampleWriteClient.startCommitWithTime(instantTime);
+      JavaRDD<WriteStatus> writeStatusRDD = sampleWriteClient.bulkInsert(jsc.parallelize(samples, 1), instantTime);
+      if (writeStatusRDD.filter(WriteStatus::hasErrors).count() > 0) {
```

Comment:
we can fetch the size from WriteStatus here itself right. why we need to let it succeed and then read the commit metadata explicitly?

Reply:
the question is: if the sample write failed, can we still rely on and use the stats? it's a design choice. i prefer it to ensure the e2e write cycle finished properly. open to discuss different options.